From a48e64fe190ecdb6d7ab852de5f768a98309bd2e Mon Sep 17 00:00:00 2001
From: Dinesh kumar <dineshkumar.varadarajan@adlinktech.com>
Date: Fri, 5 Apr 2019 14:41:49 +0530
Subject: [PATCH 21/47] Adapt LVDS and parallel RGB panel interfaces.

- Drop mxcfb0 (LVDS1), causing the remaining 3 framebuffers to get
  renumbered.
- We have a GPIO gating display panel power. Unfortunately Freescale's
  display drivers don't support regulators yet.
- Use 24 bit mode on the (physical) parallel RGB interface.
- Also switch the framebuffer (RAM) data format for LVDS and parallel RGB
  to 24 bit. Only when all framebuffers have the same format can X11
  (Xinerama) merge the screens.
- Change dcic2 (Display Content Integrity Checker) to watch LVDS0 instead
  of LVDS1.
- Make LVDS0 the primary port and remove LVDS1 from the LDB (LVDS Display
  Bridge) configurations.
- Move fb 'status' to the common .dtsi, where all devices are enabled. This
  keeps things simple, particularly for Quad/Dual.
- Only Solo/DualLite must disable either LVDS or parallel RGB devices, so
  create a specific option in the .dts.

Signed-off-by: Dinesh kumar <dineshkumar.varadarajan@adlinktech.com>
---
 arch/arm/boot/dts/lec-imx6.dtsi | 90 +++++++++++++++++------------------------
 arch/arm/boot/dts/lec-imx6q.dts | 22 +---------
 2 files changed, 37 insertions(+), 75 deletions(-)

diff --git a/arch/arm/boot/dts/lec-imx6.dtsi b/arch/arm/boot/dts/lec-imx6.dtsi
index 809ccbf..19605cc 100644
--- a/arch/arm/boot/dts/lec-imx6.dtsi
+++ b/arch/arm/boot/dts/lec-imx6.dtsi
@@ -26,7 +26,6 @@
 		mxcfb0 = &mxcfb1;
 		mxcfb1 = &mxcfb2;
 		mxcfb2 = &mxcfb3;
-		mxcfb3 = &mxcfb4;
 	};
 
 	hannstar_cabc {
@@ -52,6 +51,16 @@
 		#address-cells = <1>;
 		#size-cells = <0>;
 
+		reg_smarc_lcd_vdd: regulator@130 {
+			compatible = "regulator-fixed";
+			reg = <130>;
+			regulator-name = "smarc_lcd_vdd";
+			gpio = <&gpio1 30 GPIO_ACTIVE_HIGH>; /* SMARC LCD_VDD_EN */
+			enable-active-high;
+			regulator-always-on; /* Freescale's fb/lcd/ldb drivers can't control a regulator yet :-( */
+			regulator-state-mem { regulator-off-in-suspend; }; /* no effect? */
+		};
+
 		reg_usb_otg_vbus: regulator@0 {
 			compatible = "regulator-fixed";
 			reg = <0>;
@@ -185,17 +194,7 @@
 		hdmi-controller = <&hdmi_audio>;
 	};
 
-	mxcfb1: fb@0 {
-		compatible = "fsl,mxc_sdc_fb";
-		disp_dev = "ldb";
-		interface_pix_fmt = "RGB666";
-		default_bpp = <16>;
-		int_clk = <0>;
-		late_init = <0>;
-		status = "disabled";
-	};
-
-	mxcfb2: fb@1 {
+	mxcfb1: fb@0 {	/* IPU1 DI1: /dev/fb0=HDMI, fb1=IPU1's overlay */
 		compatible = "fsl,mxc_sdc_fb";
 		disp_dev = "hdmi";
 		interface_pix_fmt = "RGB24";
@@ -203,37 +202,39 @@
 		default_bpp = <24>;
 		int_clk = <0>;
 		late_init = <0>;
-		status = "disabled";
+		status = "okay";
 	};
 
-	mxcfb3: fb@2 {
+	mxcfb2: fb@1 {	/* IPU1 DI0: /dev/fb2=parallel RGB */
 		compatible = "fsl,mxc_sdc_fb";
 		disp_dev = "lcd";
-		interface_pix_fmt = "RGB565";
+		interface_pix_fmt = "RGB24";
 		mode_str ="CLAA-WVGA";
-		default_bpp = <16>;
+		default_bpp = <24>;
 		int_clk = <0>;
 		late_init = <0>;
-		status = "disabled";
+		status = "okay";
 	};
 
-	mxcfb4: fb@3 {
+	mxcfb3: fb@2 {	/* IPU2 DI0: /dev/fb3=LVDS, fb4=IPU2's overlay
+		         * Solo/DL: IPU1, fb2 competing with par. RGB, no overlay */
 		compatible = "fsl,mxc_sdc_fb";
 		disp_dev = "ldb";
-		interface_pix_fmt = "RGB666";
-		default_bpp = <16>;
+		interface_pix_fmt = "RGB24";
+		default_bpp = <24>;
 		int_clk = <0>;
 		late_init = <0>;
-		status = "disabled";
+		status = "okay";
 	};
 
-	lcd@0 {
+	rgb: lcd@0 {	/* parallel RGB (see fb above) */
 		compatible = "fsl,lcd";
 		ipu_id = <0>;
 		disp_id = <0>;
-		default_ifmt = "RGB565";
+		default_ifmt = "RGB24";
 		pinctrl-names = "default";
 		pinctrl-0 = <&pinctrl_ipu1>;
+		disp-power-on-supply = <&reg_smarc_lcd_vdd>; /* unsupported by driver */
 		status = "okay";
 	};
 
@@ -311,8 +312,8 @@
 };
 
 &clks {
-	assigned-clocks = <&clks IMX6QDL_CLK_LDB_DI0_SEL>,
-			  <&clks IMX6QDL_CLK_LDB_DI1_SEL>;
+	assigned-clocks = <&clks IMX6QDL_CLK_LDB_DI0_SEL>,  /*LVDS*/
+			  <&clks IMX6QDL_CLK_LDB_DI1_SEL>;  /*unused*/
 	assigned-clock-parents = <&clks IMX6QDL_CLK_PLL2_PFD0_352M>,
 				 <&clks IMX6QDL_CLK_PLL2_PFD0_352M>;
 };
@@ -357,7 +358,7 @@
 
 &dcic2 {
 	dcic_id = <1>;
-	dcic_mux = "dcic-lvds1";
+	dcic_mux = "dcic-lvds0";
 	status = "okay";
 };
 
@@ -678,8 +679,9 @@
 			>;
 		};
 
-		pinctrl_audmux: audmuxgrp {
+		pinctrl_regulators: regulatorsgrp {
 			fsl,pins = <
+				MX6QDL_PAD_ENET_TXD0__GPIO1_IO30     0x80000000 /* SMARC LCD_VDD_EN */
 				MX6QDL_PAD_CSI0_DAT7__AUD3_RXD		0x130b0
 				MX6QDL_PAD_CSI0_DAT4__AUD3_TXC		0x130b0
 				MX6QDL_PAD_CSI0_DAT5__AUD3_TXD		0x110b0
@@ -969,42 +971,22 @@
 	};
 };
 
-&ldb {
+&ldb {		/* LVDS (see fb above) */
+	disp-power-on-supply = <&reg_smarc_lcd_vdd>; /* unsupported by driver */
 	status = "okay";
 
 	lvds-channel@0 {
-		fsl,data-mapping = "spwg";
-		fsl,data-width = <18>;
+		fsl,data-mapping = "jeida"; /* spwg or jeida (see chapter "Bit Mapping" in i.MX6 Reference Manual) */
+		fsl,data-width = <24>;
+		primary;
 		status = "okay";
 
 		display-timings {
 			native-mode = <&timing0>;
 			timing0: hsd100pxn1 {
 				clock-frequency = <65000000>;
-				hactive = <1024>;
-				vactive = <768>;
-				hback-porch = <220>;
-				hfront-porch = <40>;
-				vback-porch = <21>;
-				vfront-porch = <7>;
-				hsync-len = <60>;
-				vsync-len = <10>;
-			};
-		};
-	};
-
-	lvds-channel@1 {
-		fsl,data-mapping = "spwg";
-		fsl,data-width = <18>;
-		primary;
-		status = "okay";
-
-		display-timings {
-			native-mode = <&timing1>;
-			timing1: hsd100pxn1 {
-				clock-frequency = <65000000>;
-				hactive = <1024>;
-				vactive = <768>;
+				hactive = <1024>;
+				vactive = <600>;
 				hback-porch = <220>;
 				hfront-porch = <40>;
 				vback-porch = <21>;
diff --git a/arch/arm/boot/dts/lec-imx6q.dts b/arch/arm/boot/dts/lec-imx6q.dts
index 95d76ba..c035ec1 100644
--- a/arch/arm/boot/dts/lec-imx6q.dts
+++ b/arch/arm/boot/dts/lec-imx6q.dts
@@ -20,30 +20,10 @@
 	compatible = "fsl,imx6q-sabresd", "fsl,imx6q";
 };
 
-&ldb {
+&ldb {		/* LVDS (see fb in lec-imx6.dtsi) */
 	lvds-channel@0 {
 		crtc = "ipu2-di0";
 	};
-
-	lvds-channel@1 {
-		crtc = "ipu2-di1";
-	};
-};
-
-&mxcfb1 {
-	status = "okay";
-};
-
-&mxcfb2 {
-	status = "okay";
-};
-
-&mxcfb3 {
-	status = "okay";
-};
-
-&mxcfb4 {
-	status = "okay";
 };
 
 &sata {
-- 
2.7.4

